### PR TITLE
merge subscription integration tests

### PR DESCRIPTION
**Pull request by ermalkaleci**

### CODE REVIEW

File: src/integration_tests/merge_subscription.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                    ["0x02", "bye"]
+                ]
+            })
+        );
+
+        // first subscription will unsubscribe but it shouldn't affect second subscription
+        first_sub.unsubscribe().await.unwrap();
+    });
+
+    tokio::time::sleep(std::time::Duration::from_millis(1_500)).await;
```
Comment (xlc) [inline src/integration_tests/merge_subscription.rs:167]: why 1.5s wait? I want the tests ran as fast as possible so if we need to make it sleep, it should be minimal time

File: src/integration_tests/merge_subscription.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        run_sink_tasks(&head_sub_rx, vec![SinkTask::Send(json!(1))]).await;
+        run_sink_tasks(&finalized_sub_rx, vec![SinkTask::Send(json!(1))]).await;
+    });
+
+    let config = Config {
+        endpoints: vec![format!("ws://{addr}")],
+        stale_timeout_seconds: 0,
+        server: ServerConfig {
+            listen_address: "0.0.0.0".to_string(),
+            port: 8888,
```
Comment (xlc) [inline src/integration_tests/merge_subscription.rs:47]:
```suggestion
            port: 0,
```

always specify port `0` for tests

File: src/integration_tests/merge_subscription.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    .await
+    .unwrap();
+
+    let client = Client::new(&[format!("ws://{addr}")]).await.unwrap();
+    let mut first_sub = client
+        .subscribe(subscribe_mock, vec![], unsubscribe_mock)
+        .await
+        .unwrap();
+
+    tokio::spawn(async move {
```
Comment (xlc) [inline src/integration_tests/merge_subscription.rs:89]: take the handle and unwrap it at end of tests to ensure it completes

**Review by xlc — COMMENTED** [review]
need a test to ensure upstream subscription is correctly unsubscribed when all downstream subscriptions are unsubscribed

File: src/integration_tests/merge_subscription.rs (hunk 1 above)
Comment (ermalkaleci) [reply src/integration_tests/merge_subscription.rs:167]: I'm sending msg every seconds so 1.5s is just before 3rd msg send. Probably need to reduce time to millis

**Comment by ermalkaleci** [comment]
> need a test to ensure upstream subscription is correctly unsubscribed when all downstream subscriptions are unsubscribed

This isn't that easy without some changes. Upstream subscription stays alive for at least 60s. Maybe we need to make that configurable

**Comment by xlc** [comment]
We can add a method to trigger the subscription cleanup code

**Comment by xlc** [comment]
I mean a Rust method to the middleware but I guess we can't access it in the integration tests so a configurable check will do for now

File: src/middleware/merge_subscription.rs [hunk 4]
```diff
@@ -71,15 +71,21 @@ type UpstreamSubscription = broadcast::Sender<SubscriptionMessage>;
 pub struct MergeSubscriptionMiddleware {
     client: Arc<Client>,
     merge_strategy: MergeStrategy,
+    keep_alive: u64,
```
Comment (xlc) [inline src/middleware/merge_subscription.rs:74]:
```suggestion
    keep_alive_seconds: u64,
```

**Pull request merged**